Tests for correct Pre/Semi/Metric subtyping

richardreeve opened the pull request:
I've noticed while working on #58 that there is a fairly ad hoc testing of whether the subtyping is correct for the distance measures defined here - there is much testing of equality to zero for self-distances, but not symmetry (for `SemiMetric`s) and the triangle inequality (for `Metric`s). I've introduced a test `@test_metricity`, that checks what each distance is a subtype of and tests the appropriate conditions.

There is now a systematic test for each distance measure to check it does not overclaim it's "metricity" (it can't check for underclaiming, but that is a less serious problem). This exposed a series of problems to do with floating point rounding errors for `CorrDist`, `CosineDist` and `RenyiDivergence`. This kind of issue has been identified before (e.g. #13), which I have now fixed (for `RenyiDivergence`), or relaxed the zero equality constraints (for the others).

I then updated the benchmarks and cleaned up a lot of the other testing code to remove unnecessary tests and move to a standard Julia style as seen in the rest of the package (e.g. 1. -> 1.0). Is this useful to people?

richardreeve commented:
@ararslan This was intended to fix the issue I addressed in #58 that metricity is not checked and can have implications for how things are calculated... does it make sense to you?

ararslan commented:
It does indeed, thanks. This looks good to me overall, but you've used 2 space indents in several places--if you could change those to 4 for consistency with the rest of the package, that would be great.

richardreeve commented:
Ah, thanks for pointing that out. Now fixed.

ararslan commented on src/metrics.jl:
```diff
@@ -59,14 +59,14 @@ type SpanNormDist <: SemiMetric end
 const UnionMetrics = Union{Euclidean, SqEuclidean, Chebyshev, Cityblock, Minkowski, Hamming, Jaccard, RogersTanimoto, CosineDist, CorrDist, ChiSqDist, KLDivergence, RenyiDivergence, JSDivergence, SpanNormDist}
 
 """
-    Euclidean([thresh])
```
This kind of indentation in the docstring is correct and shouldn't be removed

ararslan commented on test/test_dists.jl:
```diff
@@ -481,17 +481,17 @@ Q = Q * Q'  # make sure Q is positive-definite
-    pd = pairwise(SqEuclidean(1e-12), X)
-    @test pd[1,1] == 0
-    @test pd[2,2] == 0
+X = [0.1 0.2; 0.3 0.4; -0.1 -0.1]
```
The contents of `@testset`s should be indented, so this was correct

richardreeve replied:
Ah, all of the other `@testset`s were the other way. I've corrected them all now

ararslan commented on src/metrics.jl:
Should should be either reindented or wrapped in double backticks

ararslan commented on test/test_dists.jl:
```diff
@@ -1,68 +1,153 @@
+        dxy = evaluate(dist, x, y)
+        dxz = evaluate(dist, x, z)
+        dyz = evaluate(dist, y, z)
+        if (isa(dist, PreMetric))
```
No parens around conditions

ararslan commented on test/test_dists.jl:
```diff
@@ -101,28 +182,24 @@ w = rand(size(a))
 end
 @test kl_divergence(p, q) ≈ klv
 
-@test renyi_divergence(p, p, 0) ≈ 0
```
Why are these being removed?

richardreeve replied:
Because they are all tested in the Premetric test - Premetrics are non-negative and zero when measuring a distance from something to itself.

richardreeve commented:
Thanks @ararslan - as I say in the comments, the Premetric test includes checks for equality to zero for self-distance, so they don't need to be repeated (as well as many other tests). And all but one of the `@testset`s were not indented originally, but I've now fixed them all (I think!).

ararslan commented on src/metrics.jl:
```diff
@@ -86,10 +86,10 @@ julia> pairwise(Euclidean(1e-12), x, x)
 
 """
-    SqEuclidean([thresh])
+SqEuclidean([thresh])
```
Should also be indented

richardreeve commented:
Oops, thanks.

richardreeve commented:
Small test fix required as Weighted Hamming turned out to have similar rounding errors to other metrics, resulting in test failure for dxz ≤ dxy + dyz - fixed by allowing dxz ≈ dxy + dyz for Metrics, not just strict equality.

KristofferC commented on README.md:
```diff
@@ -151,7 +151,7 @@ Each distance corresponds to a distance type. The type name and the correspondin
 |  WeightedHamming     |  `whamming(x, y, w)`       | `sum((x .!= y) .* w)`  |
 
-**Note:** The formulas above are using *Julia*'s functions. These formulas are mainly for conveying the math concepts in a concise way. The actual implementation may use a faster way.
+**Note:** The formulas above are using *Julia*'s functions. These formulas are mainly for conveying the math concepts in a concise way. The actual implementation may use a faster way. x and y arguments are arrays of real numbers; k and l arguments are arrays of distinct elements of any kind; a and b arguments are arrays of `Bool`s; and p and q arguments are arrays forming a discrete probability d ...
```
Perhaps:

"The arguments `x` and `y` are arrays of real numbers; `k` and `l` are arrays of distinct elements of any kind; `a` and `b` are arrays of `Bool`s; and finally, `p` and `q`  are arrays forming a discrete probability distribution and are therefore both expected to sum to one."

richardreeve replied:
Fixed.

KristofferC commented on src/metrics.jl:
````diff
@@ -86,7 +86,7 @@ julia> pairwise(Euclidean(1e-12), x, x)
 ```
 """
-Euclidean() = Euclidean(0)
+    Euclidean() = Euclidean(0)
````
Why indent here?

ararslan replied:
OH WOW I'M SO DUMB. I thought this was the top of the docstring when I was doing my initial review so I told @richardreeve to indent it. Sorry!

richardreeve replied:
Oops, getting carried away.

KristofferC commented on test/test_dists.jl:
```diff
@@ -1,245 +1,332 @@
+            # in our tests due to accumulating floating point rounding errors.
+            # We either need to allow small errors in our tests or change the
+            # way we do accumulations...
+            @test evaluate(dist, x, x) + one(eltype(x)) ≈ one(eltype(x))
```
Doesn't `evaluate(dist, x, x) ≈ 0` work? Otherwise perhaps use `isapprox`?

KristofferC replied:
Also, if you want, this while thing could just be changed to a function instead of a macro.

richardreeve replied:
No, it doesn't - again there are too many small errors in accumulation of the floating point numbers. The probabilities that these often represent are ≈ 1, so I thought that was a reasonable precision to use - otherwise I'd have been picking some arbitrary error tolerance using approx() that should more reasonably have depended on the floating point precision being used. You're right it could also be put into a function - in fact I had it as that earlier - but I didn't feel it added to the clarity of the code.

KristofferC commented on src/metrics.jl:
```diff
@@ -208,46 +208,46 @@ kl_divergence(a::AbstractArray, b::AbstractArray) = evaluate(KLDivergence(), a,
 
 # RenyiDivergence
 function eval_start{T<:AbstractFloat}(::RenyiDivergence, a::AbstractArray{T}, b::AbstractArray{T})
```
Should these be called `p` and `q` like in the README?

richardreeve replied:
Honestly, these argument names would have to be changed throughout the code for a lot of the distance metrics - I can do that, but it'll look like I wrote the whole package... I wasn't sure it was worth it, but I'm happy to do it if you like?

KristofferC commented on src/metrics.jl:
```diff
@@ -208,46 +208,46 @@ kl_divergence(a::AbstractArray, b::AbstractArray) = evaluate(KLDivergence(), a,
 # RenyiDivergence
 function eval_start{T<:AbstractFloat}(::RenyiDivergence, a::AbstractArray{T}, b::AbstractArray{T})
-    zero(T), zero(T)
+    zero(T), zero(T), sum(a), sum(b)
```
If `p` and `q` (`a` and `b`) should sum to 1, why do we need to compute their sum here?

richardreeve replied:
@KristofferC This is because of accumulation errors in `Float64`s - they turn out to be strewn throughout these distance measures unfortunately. sum(a) is often 0.99999999999 or so, and this causes v slight negative distances in some places, and other similar errors. I fixed them here for Renyi, but I don't know think there's a general way of doing it for other metrics, particularly when they are not probability distributions, so they shouldn't sum to 1 anyway.

richardreeve commented:
@KristofferC I've added in some docs, and addressed your issues (some with revisions, but the impossible-to-avoid problem is rounding errors which caused the others). Is this okay to merge now?

ararslan commented on src/metrics.jl:
```diff
@@ -29,6 +29,40 @@ type CorrDist <: SemiMetric end
+
+At α = 1, ``R_1(P | Q) = sum_{i: p_i > 0}(p_i log(p_i / q_i))``
+
+At α = ∞, ``R_∞(P | Q) = log(sup_{i: p_i > 0}(p_i / q_i))``
```
In math sections, Unicode literals should be replaced with the equivalent LaTeX, e.g. `\infty` in place of `∞`

richardreeve replied:
Actually, much to my surprise, the [docs](http://docs.julialang.org/en/latest/manual/documentation.html) say the opposite:

```
Use backticks to identify code and equations.

Julia identifiers and code excerpts should always appear between backticks ` to enable
highlighting. Equations in the LaTeX syntax can be inserted between double backticks ``.
Use Unicode characters rather than their LaTeX escape sequence, i.e. ``α = 1`` rather
than ``\\alpha = 1``.
```

I suspect it's because when you do `?RenyiDivergence` at the command line, it can't actually do latex, so you're better off with this bastardised version...

ararslan replied:
Wow, huh. TIL. Okay, I guess this is fine as-is, then.

richardreeve commented:
@ararslan So, can we go ahead and merge this then?

ararslan commented:
In light of https://github.com/JuliaStats/Distances.jl/pull/70, could you rebase out the commits that change the benchmarking stuff?
```git
git checkout metricity
git rebase -i HEAD~25  # since there are 25 commits here
```
Then delete commits as needed. That should reduce the amount of branch conflict resolution needed in conjunction with #70.

richardreeve commented:
@ararslan I have to admit I'm quite hazy on what I did there, but I resolved the conflicts after doing the rebase you suggested. Hopefully that helped.

ararslan commented:
Okay, that didn't quite work, but that's fine. We can just merge this and rebase #70. Does that sound reasonable to you, @KristofferC?

richardreeve commented:
@KristofferC - is this good to go then?
